src/pudl/analysis/allocate_net_gen.py

```diff
@@ -1598,6 +1596,80 @@ def adjust_energy_source_codes(
     return gens
 
 
+def add_missing_energy_source_codes_to_gens(gens_at_freq, gf, bf):
+    """Add energy_source_codes that appear in the `gf` table but not `gens` to `gens`.
+
+    In some cases, non-zero fuel consumption and net generation is reported in the
+    EIA-923 generation and fuel table that is associated with an energy_source_code that
+    is not associated with that plant-prime mover in the gens table, which would cause
+    these data to get dropped when these two tables are merged. To fix this, for each
+    plant-pm, this function identifies such esc, and adds them to the `gens_at_freq`
+    table as new energy_source_code columns.
+    """
+    missing_gf_escs_from_gens = identify_missing_gf_escs_in_gens(gens_at_freq, gf, bf)
+    if missing_gf_escs_from_gens.empty:
+        return gens_at_freq
+    idx = ["plant_id_eia", "prime_mover_code", "report_date"]
+    # pivot these data to become new numbered energy_source_code_n columns starting at 8
+    # the native table ends at 6, but we add one more in adjust_msw_energy_source_codes
+    missing_gf_escs_from_gens["num"] = (
+        missing_gf_escs_from_gens.groupby(idx).cumcount() + 8
+    )
+    missing_gf_escs_from_gens["num"] = missing_gf_escs_from_gens["num"].astype(str)
+    missing_gf_escs_from_gens = missing_gf_escs_from_gens.pivot(
+        index=idx, columns="num"
+    )[["energy_source_code"]]
+    missing_gf_escs_from_gens.columns = [
+        "_".join(col) for col in missing_gf_escs_from_gens.columns.values
+    ]
+    missing_gf_escs_from_gens = missing_gf_escs_from_gens.reset_index()
+
+    # merge these missing fuel columns into the gens data
+    gens_at_freq = gens_at_freq.merge(
+        missing_gf_escs_from_gens,
+        how="left",
+        on=idx,
+        validate="m:m",
+    )
+
+    return gens_at_freq
+
+
+def identify_missing_gf_escs_in_gens(gens_at_freq, gf, bf):
```

Comment:
Hmm it's been a while since I've looked at this so I forget exactly why I dropped the zeros, but it might have had something to do with escs associated with proposed or retired generators? What were the output metrics that improved?

Also, with the addition of the bf table and removing the zero dropping, you might need to update the docstring and inline comments.

Reply:
I could definitely imagine that a lot of these zero-data ESCs are associated with non-operating generators, but that should be covered by your `remove_inactive_generators`. I tried turning this on and off and it seemed like we were getting much more unassociated records when we dropped the nulls, so I dropped the dropping.

for metrics, I've mostly been looking at the indicators in the logs. I've kept a little running list of [the logs for the different configs here](https://github.com/catalyst-cooperative/pudl/pull/2446#issuecomment-1489339208).

These updates overall have decreased the # of records that are outside of tolerance, which is great.